cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedCall.qll

```diff
@@ -206,7 +206,44 @@ abstract class TranslatedCall extends TranslatedExpr {
 
   predicate hasPreciseSideEffect() { exists(getSideEffects()) }
 
-  TranslatedSideEffects getSideEffects() { result.getCall() = expr }
+  final TranslatedSideEffects getSideEffects() { result.getExpr() = expr }
+}
+
+abstract class TranslatedSideEffects extends TranslatedElement {
+  abstract Expr getExpr();
+
+  final override Locatable getAST() { result = getExpr() }
+
+  final override Function getFunction() { result = getExpr().getEnclosingFunction() }
+
+  override TranslatedElement getChild(int i) {
+    result =
+      rank[i + 1](TranslatedSideEffect tse, int isWrite, int index |
+        (
+          tse.getCall() = getExpr() and
```

Comment:
I don't understand why `TranslatedAllocationSideEffects` needs to inherit from this class. I presume it's for the code sharing that gives it this `getChild` predicate, but is that what we want? Should an allocation side effect have these argument-specific side effects as children? If an `AllocationExpr` is also a `Call`, won't these argument-specific side effects then have two parents?

Reply:
Indeed, it was for the code sharing, but I think you're right. I'll fix this.

Reply:
8fdc4b037abe7163e37478d86603702d15f97a94 has an easy fix that prevents having side effects with two parents. It doesn't address the design decision of whether the allocation side effect should also handle the call side effects for calls that are also allocations (or whether it should be a side effect at all). If you want me to address this in this PR feel free to say so. I think it requires a slightly more significant code change, though.